### PR TITLE
Add worker drain signal

### LINKED ISSUE #3452: ObjectDisposedException during host shutdown
### Problem Description

We are running Durable Functions orchestrations on Azure Functions Isolated Worker (.NET 10, Premium plans, Linux). When the Azure Functions platform performs a scale-in (removes a worker instance), in-flight orchestrator functions throw **`System.ObjectDisposedException: Cannot access a disposed object. Object name: 'IServiceProvider'`** and are permanently marked as `RuntimeStatus = Failed`. The Durable Task framework does **not** retry these orchestrations — they remain in `Failed` state indefinitely, causing upstream workflows to become stuck.

#### The cascade

When the host initiates shutdown for scale-in, the DI `ServiceProvider` is disposed. However, orchestrator functions that have **already been dispatched** (messages dequeued from the control queue and handed to the worker) attempt to resolve services via `DefaultFunctionContext.get_InstanceServices()` which calls `ServiceProvider.CreateScope()`. This throws `ObjectDisposedException` because the container is already disposed.

The Durable Task extension treats this as an unhandled orchestrator failure and commits `RuntimeStatus = Failed` to the Instances table. The failure then propagates up the orchestration call chain:

1. **`wait-for-task`** (sub-orchestration) — fails first with `ObjectDisposedException`
2. **`migrate-site-workflow`** (parent) — receives `SubOrchestrationInstanceFailed` from `wait-for-task` and also transitions to `Failed`
3. **`process-one-backlog-item-workflow`** (top-level) — receives the cascading failure and permanently fails

**Inconsistent retry behavior**: In some cases, the framework appears to retry the failed orchestration — the queue message is picked up by another instance and the workflow recovers. However, in other cases (the majority we observe), the failure is permanent: `RuntimeStatus = Failed` is committed to the Instances table and no retry occurs. The business workflow is permanently abandoned with no automatic recovery.

**The critical issue**: Unlike activity function failures which leave queue messages available for retry, orchestrator failures *sometimes* commit a terminal `Failed` status to the instance store. Once committed, the orchestration is permanently dead — no re-dispatch. The non-deterministic nature of retry vs permanent failure makes this particularly dangerous: there is no reliable way to predict which orchestrations will recover and which will be permanently lost.

#### Observed timeline (real production incident, 2026-06-01 13:45 UTC)

| Time (UTC) | Event |
|---|---|
| 13:45:23.834 | Host logs: _"Application is shutting down..."_ on instance `0--bb3e733a...` |
| 13:45:24.015 | 29 orchestrations fetch `TimerFired` messages from control queues |
| 13:45:24.051 | Orchestrations begin processing (timer delays 63–64 seconds) |
| 13:45:24.098 | `ObjectDisposedException` ×28 thrown — `ServiceProvider` already disposed |
| 13:45:24.155 | `wait-for-task` orchestrations set to `RuntimeStatus: Failed` |
| 13:45:24.165 | `SubOrchestrationInstanceFailed` messages sent to parent orchestrations |
| 13:45:24.197 | Instance tables updated: 29 orchestrations permanently `Failed` |
| 13:45:28.445 | Parent `process-one-backlog-item-workflow` orchestrations also `Failed` |

After shutdown, the instance **never appeared again** — confirming this was a permanent scale-in, not a restart.

#### Key observations

1. **The ServiceProvider is disposed before in-flight invocations complete.** The shutdown signal and ServiceProvider disposal happen within 28–478ms of each other. Orchestrator functions already dispatched cannot resolve dependencies.

2. **Orchestrator failures are permanent — no retry.** Unlike activity failures (which return the message to the queue), when an orchestrator function fails, the extension commits `RuntimeStatus = Failed` to the Instances table. The Durable Task framework has no built-in retry for orchestrator failures.

3. **One scale-in event kills many orchestrations.** In the worst observed incident, a single instance shutdown caused **29 sub-orchestrations** and their parent workflows to fail permanently. With many orchestrations sharing a single instance, one scale-in can cascade into dozens of permanently abandoned workflows.

4. **SubOrchestrationInstanceFailed propagates fatally.** When a child `wait-for-task` orchestration fails, the parent `process-one-backlog-item-workflow` receives `SubOrchestrationInstanceFailed` and also transitions to `Failed` — extending the blast radius beyond the originally affected instance.

5. **Contrast: Timer-based orchestrations survive.** Orchestrations that were merely waiting on timers (not actively executing) survive because their timer messages remain on the control queue and are picked up by a new instance. Only orchestrations whose code was **actively executing** at shutdown time are permanently lost.

#### Impact (3-day observation window, EU region only)

| Metric | Value |
|---|---|
| Total `ObjectDisposedException` occurrences | 169 |
| Unique worker instances affected (scale-in events) | 7 |
| Permanently failed orchestration instances | 39 |
| Affected function apps | `odmt-eu-b1-files-workers`, `odmt-eu-b1-mail-workers` |
| Failed orchestration types | `wait-for-task`, `process-one-backlog-item-workflow` |
| Customer impact | Migration tasks stuck permanently — no completion |

#### Extension log confirming permanent failure

```
739294d1f4be5a8ab7541cc46f088164: Function 'wait-for-task (Orchestrator)' failed with an error.
Reason: Microsoft.Azure.WebJobs.Host.FunctionInvocationException
IsReplay: False. State: Failed. RuntimeStatus: Failed.
HubName: filesworkers. AppName: odmt-eu-b1-files-workers.
SlotName: Production. ExtensionVersion: 3.12.5. SequenceNumber: 16420. TaskEventId: -1
```

#### Exception stack trace

```
Microsoft.Azure.WebJobs.Host.FunctionInvocationException:
  Exception while executing function: Functions.process-one-backlog-item-workflow
 ---> Microsoft.Azure.WebJobs.Script.Workers.Rpc.RpcException:
  Result: Failure
  Type: System.ObjectDisposedException
  Exception: Cannot access a disposed object.
  Object name: 'IServiceProvider'.
  Stack:
    at Microsoft.Extensions.DependencyInjection.ServiceLookup.ThrowHelper
       .ThrowObjectDisposedException()
    at Microsoft.Extensions.DependencyInjection.ServiceProvider.CreateScope()
    at Microsoft.Azure.Functions.Worker.DefaultFunctionContext
       .get_InstanceServices()
       in /*/src/DotNetWorker.Core/Context/DefaultFunctionContext.cs:line 46
    at Microsoft.Extensions.Hosting.MiddlewareWorkerApplicationBuilderExtensions
       .<>c__DisplayClass3_0`1.b__1(FunctionContext context)
       in /_/src/DotNetWorker.Core/Hosting/WorkerMiddlewareWorkerApplicationBuilderExtensions.cs:line 105
    at Microsoft.Azure.Functions.Worker.FunctionsApplication
       .InvokeFunctionAsync(FunctionContext context)
       in /_/src/DotNetWorker.Core/FunctionsApplication.cs:line 83
    at Microsoft.Azure.Functions.Worker.Handlers.InvocationHandler
       .InvokeAsync(InvocationRequest request)
       in /_/src/DotNetWorker.Grpc/Handlers/InvocationHandler.cs:line 91
```

#### Expected behavior

When a Functions host undergoes shutdown/scale-in:
1. Pending orchestrator messages should be abandoned (left on queue) so another instance picks them up
2. If an orchestrator function invocation fails due to host disposal, the failure should be treated as retriable (not permanently terminal)
3. The `ServiceProvider` disposal should be deferred until all in-flight function invocations complete (graceful drain)

---

**Product**: Durable Azure Functions (Isolated Worker, .NET 10)  
**Plan**: Flex Consumption
**Durable Task Extension**: `Microsoft.Azure.Functions.Worker.Extensions.DurableTask 1.16.5`    
**Task Hub**: `filesworkers`, `mailboxworkers`
**OS**: Linux  
**Regions affected**: North Europe  

---

### PR DESCRIPTION
**Pull request by sophiatev**
# Summary
## What changed?
This PR introduces a new worker drain signal in the case of Function execution - if this signal is set and the Function failed, we will retry the execution rather than permanently failing it.

## Why is this change needed?
We have this signal for host shutdown but this is decoupled from worker shutdown. 

## Issues / work items
- Resolves #3452 

# Project checklist
- [ ] Documentation changes are not required
  - [ ] Otherwise: Documentation PR is ready to merge and referenced in `pending_docs.md`
- [ ] Release notes are not required for the next release
  - [ ] Otherwise: Notes added to `release_notes.md`
- [ ] Backport is not required
  - [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
- [ ] All required tests have been added/updated (unit tests, E2E tests)
- [ ] No extra work is required to be leveraged by OutOfProc SDKs
  - [ ] Otherwise: Work tracked here: #issue_or_pr_in_each_sdk
- [ ] No change to the version of the `WebJobs.Extensions.DurableTask` package
  - [ ] Otherwise: Major/minor updates are reflected in `/src/Worker.Extensions.DurableTask/AssemblyInfo.cs`
- [ ] No EventIds were added to `EventSource` logs
  - [ ] Otherwise: Ensure EventIds are within the supported range in the existing Windows infrastructure (validate via deployed telemetry). If needed, extend the range via a PR such as https://msazure.visualstudio.com/One/_git/AAPT-Antares-Websites/pullrequest/7463263?_a=files
- [ ] This change should be added to the `v2.x` branch
  - [ ] Otherwise: This change applies exclusively to `WebJobs.Extensions.DurableTask` v3.x and will be retained only in the `dev` and `main` branches
- [ ] Breaking change?
  - [ ] If yes:
    - Impact:
    - Migration guidance:
---

# AI-assisted code disclosure (required)
## Was an AI tool used? (select one)
- [ ] No
- [ ] Yes, AI helped write parts of this PR (e.g., GitHub Copilot)
- [ ] Yes, an AI agent generated most of this PR

If AI was used:
- Tool(s):
- AI-assisted areas/files:
- What you changed after AI output:

AI verification (required if AI was used):
- [ ] I understand the code and can explain it
- [ ] I verified referenced APIs/types exist and are correct
- [ ] I reviewed edge cases/failure paths (timeouts, retries, cancellation, exceptions)
- [ ] I reviewed concurrency/async behavior
- [ ] I checked for unintended breaking or behavior changes

---

# Testing
## Automated tests
- Result: Passed / Failed (link logs if failed)

## Manual validation (only if runtime/behavior changed)
- Environment (OS, .NET version, components):
- Steps + observed results:
  1.
  2.
  3.
- Evidence (optional):

---

# Notes for reviewers
- N/A

### CODE REVIEW

File: src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs [hunk 1]
```diff
@@ -177,9 +180,15 @@ await this.LifeCycleNotificationHelper.OrchestratorStartingAsync(
...
                     // Re-throw so we can abort this invocation.
-                    this.HostLifetimeService.OnStopping.ThrowIfCancellationRequested();
-                }
-
+                    this.HostLifetimeService.OnStopping.ThrowIfCancellationRequested();
+
+                    // The Function may have failed because the worker is draining, so we want to retry its execution in this case
+                    if (isWorkerDraining)
+                    {
+                        throw new Exception("Worker is shutting down. The orchestration execution will be aborted and retried.");
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs:188]: We shouldn't be throwing `System.Exception` as a general practice. Why not `SessionAbortedException` like we do elsewhere?

File: src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs [hunk 2]
```diff
@@ -418,6 +431,12 @@ void SetErrorResult(FailureDetails failureDetails)
                         throw functionResult.Exception;
                     }
 
+                    // The Function may have failed because the worker is draining, so we want to retry its execution in this case
+                    if (isWorkerDraining)
+                    {
+                        throw new Exception("Worker is shutting down. The entity execution will be aborted and retried.");
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs:437]: Same comment about the type of exception we throw.

File: src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs [hunk 3]
```diff
@@ -596,6 +617,19 @@ public async Task CallActivityAsync(DispatchMiddlewareContext dispatchContext, F
...
+            }
+
+            // The Function may have failed because the worker is draining, so we want to retry its execution in this case
+            if (!result.Succeeded && IsWorkerDrainingException(result.Exception))
+            {
+                this.TraceHelper.FunctionAborted(
+                    this.Options.HubName,
+                    functionName.Name,
+                    instance.InstanceId,
+                    "The worker is shutting down. The activity execution will be aborted and retried.",
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs:629]: Consider saying "out-of-process worker" instead of just "worker" in all of these messages to disambiguate from the _durable task_ worker.

File: src/Worker.Extensions.DurableTask/Exceptions/WorkerDrainingException.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+internal sealed class WorkerDrainingException : Exception
+{
+    /// <summary>
+    /// Stable marker the host matches on to detect this exception after it crosses the gRPC boundary
+    /// as a serialized string. The host keeps an identical copy of this literal; keep them in sync.
+    /// </summary>
+    internal const string Sentinel = "[DurableTask:WorkerDraining]";
+
+    public WorkerDrainingException()
+        : base($"The worker is shutting down and will not commit this result; the work item should be retried on another worker. {Sentinel}")
```
Comment (cgillum) [inline src/Worker.Extensions.DurableTask/Exceptions/WorkerDrainingException.cs:21]: nit: it probably makes sense to put the sentinel in the beginning of the message. That also allows us to do a more efficient `StartsWith` string comparison instead of a `Contains`.

File: src/Worker.Extensions.DurableTask/Execution/DurableFunctionExecutor.Activity.cs [hunk 5]
```diff
@@ -28,11 +28,16 @@ private async ValueTask RunActivityAsync(FunctionContext context, BindingMetadat
             if (context.FunctionDefinition.EntryPoint == ActivityEntryPoint)
             {
                 await this.RunDirectActivityAsync(context, triggerBinding);
-                return;
+            }
+            else
```
Comment (cgillum) [inline src/Worker.Extensions.DurableTask/Execution/DurableFunctionExecutor.Activity.cs:32]: What is this `else` branch change for? It seems to be adding unnecessary complexity to the diff and the code overall.

File: src/Worker.Extensions.DurableTask/Execution/DurableFunctionExecutor.cs [hunk 6]
```diff
@@ -15,11 +16,15 @@ internal partial class DurableFunctionExecutor(
...
     IDurableTaskFactory factory,
     IOptions<DurableTaskWorkerOptions> options,
+    IHostApplicationLifetime? hostApplicationLifetime = null,
     IExceptionPropertiesProvider? exceptionPropertiesProvider = null)
     : IFunctionExecutor
 {
     private DataConverter Converter => options.Value.DataConverter;
 
+    private bool IsWorkerDraining =>
+        hostApplicationLifetime?.ApplicationStopping.IsCancellationRequested ?? false;
```
Comment (cgillum) [inline src/Worker.Extensions.DurableTask/Execution/DurableFunctionExecutor.cs:26]: The `DurableFunctionsExecutor` runs really late in the execution pipeline so it will only work in cases where execution actually makes it here. If I understand correctly, the person who opened the original issue observed an exception that happened somewhere else in the pipeline, suggesting that cancellation-related issues can occur before or after we get to this middleware. So, it seems that this will catch some cases but not all. A more robust approach would be to add this kind of logic directly in the .NET Isolated worker code rather than trying to put it in the Durable Functions worker extension code.
